File: src/main/java/dev/qinx/faker/annotation/person/Name.java
package dev.qinx.faker.annotation.person;

import dev.qinx.faker.enums.Gender;
import dev.qinx.faker.enums.Locale;
import dev.qinx.faker.provider.Provider;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.FIELD})
public @interface Name {
    Locale locale() default Locale.en;
    Gender gender() default Gender.All;
    String firstName() default "true";
    String lastName() default "true";
    String seed() default "";
    Class<? extends Provider> provider() default dev.qinx.faker.provider.person.NameProvider.class;
}

File: src/main/scala/dev/qinx/faker/provider/person/Name.scala
package dev.qinx.faker.provider.person

import java.lang.annotation.Annotation

import dev.qinx.faker.enums.Locale
import dev.qinx.faker.internal.{HasRandom, Logging}
import dev.qinx.faker.provider.Provider
import dev.qinx.faker.utils.ReflectUtils

class Name extends Provider[String] with HasRandom with Logging {

  private[this] var provider: LocalNameProvider = _

  override def provide(): String = this.provider.provide()

  override def setSeed(seed: Long): Name.this.type = {
    this.provider.setSeed(seed)
    this
  }

  override def configure(annotation: Annotation): this.type = {
    val locale = ReflectUtils.invokeAnnotationMethod[Locale](annotation, "locale")
    this.provider = Name(locale)
    this.provider.configure(annotation)
    this
  }

}

object Name {

  def apply(locale: Locale): LocalNameProvider = {
    val providerName = s"dev.qinx.faker.provider.person.${locale.name()}.NameProvider"

    Class
      .forName(providerName)
      .getDeclaredConstructors
      .head
      .newInstance()
      .asInstanceOf[LocalNameProvider]
  }

}
